internal/dag/httpproxy_processor.go

```diff
@@ -1381,17 +1380,21 @@ func toStringSlice(hvs []contour_api_v1.CORSHeaderValue) []string {
 }
 
 func includeMatchConditionsIdentical(includes []contour_api_v1.Include) bool {
-	j := 0
-	for i := 1; i < len(includes); i++ {
-		// Now compare each include's set of conditions
-		for _, cA := range includes[i].Conditions {
-			for _, cB := range includes[j].Conditions {
-				if (cA.Prefix == cB.Prefix) && equality.Semantic.DeepEqual(cA.Header, cB.Header) {
-					return true
-				}
-			}
+	seenIncludeConditions := map[string]struct{}{}
```

Comment:
I was never crazy about our implementation of converting sets of conditions to strings to check for equality. I think we'd probably be able to come up with some adversarial inputs that would cause it to fail, since we're not using "safe" delimiters. But, practically speaking, it probably gets the job done for most normal inputs. I don't have an issue with using the DAG structs here, better than having to duplicate the stringification logic.

Reply:
another option after working on this for a little, still messy but no string comparisons:

```go
	seenConds := map[string][][]HeaderMatchCondition{}
	for _, include := range includes {
		pathPrefix := mergePathMatchConditions(include.Conditions).Prefix

		includeHeaderConds := mergeHeaderMatchConditions(include.Conditions)
		sort.SliceStable(includeHeaderConds, func(i, j int) bool {
			if includeHeaderConds[i].MatchType != includeHeaderConds[j].MatchType {
				return includeHeaderConds[i].MatchType < includeHeaderConds[j].MatchType
			}
			if includeHeaderConds[i].Invert != includeHeaderConds[j].Invert {
				return includeHeaderConds[i].Invert
			}
			if includeHeaderConds[i].Name != includeHeaderConds[j].Name {
				return includeHeaderConds[i].Name < includeHeaderConds[j].Name
			}
			if includeHeaderConds[i].Value != includeHeaderConds[j].Value {
				return includeHeaderConds[i].Value < includeHeaderConds[j].Value
			}
			return false
		})

		seenHeaderConds, pathSeen := seenConds[pathPrefix]
		if !pathSeen {
			seenConds[pathPrefix] = [][]HeaderMatchCondition{
				includeHeaderConds,
			}
			continue
		}

		for _, headerConds := range seenHeaderConds {
			if len(headerConds) != len(includeHeaderConds) {
				seenConds[pathPrefix] = append(seenConds[pathPrefix], includeHeaderConds)
				continue
			}

			headerCondsIdentical := true
			for i := range headerConds {
				if headerConds[i] != includeHeaderConds[i] {
					seenConds[pathPrefix] = append(seenConds[pathPrefix], includeHeaderConds)
					headerCondsIdentical = false
					break
				}
			}
			if headerCondsIdentical {
				return true
			}
		}
	}

	return false
```

Reply:
a little better than comparing each condition set for an include to each and every other one, but the header conditions comparison still has to collect all unique sets and compare them